packages/maskbook/src/plugins/Trader/trader/uniswap/useAllCommonPairs.ts

```diff
@@ -20,49 +20,47 @@ export function useAllCurrencyCombinations(currencyA?: Currency, currencyB?: Cur
         const additionalA = tokenA ? context?.ADDITIONAL_TOKENS[chainId]?.[tokenA.address] ?? [] : []
         const additionalB = tokenB ? context?.ADDITIONAL_TOKENS[chainId]?.[tokenB.address] ?? [] : []
         return [...common, ...additionalA, ...additionalB].map((x) => toUniswapToken(chainId, x))
-    }, [chainId, chainIdValid, tokenA, tokenB])
+    }, [chainId, chainIdValid, tokenA?.address, tokenB?.address])
 
     const basePairs: [Token, Token][] = useMemo(
         () => flatMap(bases, (base): [Token, Token][] => bases.map((otherBase) => [base, otherBase])),
         [bases],
     )
 
-    return useMemo(
-        () =>
-            tokenA && tokenB
-                ? [
-                      // the direct pair
-                      [tokenA, tokenB],
-                      // token A against all bases
-                      ...bases.map((base): [Token, Token] => [tokenA, base]),
-                      // token B against all bases
-                      ...bases.map((base): [Token, Token] => [tokenB, base]),
-                      // each base against all bases
-                      ...basePairs,
-                  ]
-                      .filter((tokens): tokens is [Token, Token] => Boolean(tokens[0] && tokens[1]))
-                      .filter(([t0, t1]) => t0.address !== t1.address)
-                      .filter(([tokenA, tokenB]) => {
-                          if (!chainIdValid) return true
-                          const customBases = context?.CUSTOM_TOKENS?.[chainId]
+    return useMemo(() => {
+        return tokenA && tokenB
+            ? [
+                  // the direct pair
+                  [tokenA, tokenB],
+                  // token A against all bases
```

Comment:
_Change it or not_

This is a loong expression. To make it more clear, we could break it into two ones.

```ts
if (!tokenA || !tokenB) return
return [...].filter().filter()
```

The minified results would be the same.